### DISCUSSION TITLE
Adding a label via web API

### DISCUSSION THREAD

**Discussion by insomniacslk** [photoprism/photoprism#3777] [General]
Hi, I am trying to add a label to a photo using the web API. I couldn't find the API-specific docs so I looked at the source code, which led me to partial results: the label is added to the correct photo, the uncertainty value is correct, but the name is "Unknown", like if an empty string was passed.  I appreciate any pointers to address this.

The command to reproduce what I have so far is
```
curl -X POST -v \
    https://example.org/api/v1/photos/<photo_id>/label \
    -H "Content-Type: application/json" \
    -d '{"Uncertainty": 5, "Label":{"Name": "test"}}' \
    -H "X-Session-ID: 1234"
```

Thanks in advance!

**Comment by craiga** [comment]
@insomniacslk After banging my head against a wall for longer than I'd like to admit, I finally figured this out.

You need to send "Name", "Priority", and "Uncertainty" in the payload, and uncertainty seems to need to be an integer. Here's my Python which is doing this right now:

```python
        for label, confidence in labels:
            uncertainty = int(100 - confidence)
            logger.info(f"Adding label {label} with uncertainty of {uncertainty}.")
            label_response = session.post(
                api_url + "/photos/" + search_result["UID"] + "/label",
                json={
                    "Name": label,
                    "Priority": 0,
                    "Uncertainty": uncertainty
                },
            )
            label_response.raise_for_status()
```

**Reply by lastzero** [reply]
My apologies for not having answered this question! Our API generally expects all fields to be present. So you should submit the same fields (with the same value types) as our web interface. We have started to improve our API docs, but it will take some time until all endpoints and use cases are fully covered, e.g:
- https://docs.photoprism.app/developer-guide/api/
- https://docs.photoprism.app/developer-guide/api/search/
- https://docs.photoprism.app/developer-guide/api/thumbnails/#image-endpoint-uri
